share/shwap/p2p/bitswap/block_fetch.go

```diff
@@ -0,0 +1,143 @@
+package bitswap
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"sync"
+
+	"github.com/ipfs/boxo/exchange"
+	"github.com/ipfs/go-cid"
+
+	"github.com/celestiaorg/celestia-node/share"
+)
+
+// Fetch fetches and populates given Blocks using Fetcher wrapping Bitswap.
+//
+// Validates Block against the given Root and skips Blocks that are already populated.
+// Gracefully synchronize identical Blocks requested simultaneously.
+// Blocks until either context is canceled or all Blocks are fetched and populated.
+func Fetch(ctx context.Context, fetcher exchange.Fetcher, root *share.Root, blks ...Block) error {
+	cids := make([]cid.Cid, 0, len(blks))
+	duplicate := make(map[cid.Cid]Block)
+	for _, blk := range blks {
+		if !blk.IsEmpty() {
+			continue // skip populated Blocks
+		}
+
+		idStr := blk.String()
+		p := blk.Populate(root)
+		cid := blk.CID()
+		cids = append(cids, cid)
+		// store the PopulateFn s.t. hasher can access it
+		// and fill in the Block
+		_, exists := populators.LoadOrStore(idStr, p)
```

Comment:
Why we use `idStr` (`blk.String()`) instead of `cid` (`blk.CID()`) ?

Reply:
Because CID adds some additional metadata to the ID body(codec, multihash code, version) we would have to write logic that adds those on the verification side in `hasher.Write`

Reply:
But honestly, as that is a bitswap composition, we could pass CIDs over the wire instead of IDs. Yes, that would add 4 byte overhead to every message, but would simplify protocol implementations.

Reply:
cc @walldiss for thoughts

Reply:
Ok, I tried that, and I like that more than before, so I am sticking to CID over the wire